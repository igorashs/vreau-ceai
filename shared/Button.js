import styled, { css } from 'styled-components';

export const buttonStyle = css`
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  text-decoration: none;

  cursor: pointer;

  :hover {
    background-color: var(--accent);
  }
`;

const noneStyle = css`
  background-color: transparent;
  box-shadow: none;

  :hover {
    background-color: #0000001a;
  }
`;

const dangerStyle = css`
  background-color: var(--danger-dark);

  :hover {
    background-color: var(--danger);
  }
`;

const dangerTextStyle = css`
  color: var(--text-danger-light);
  background-color: transparent;
  box-shadow: none;

  :hover {
    background-color: transparent;
    filter: brightness(0.7);
  }
`;

export const Button = styled.button`
  display: flex;
  align-items: center;
  gap: 7px;

  ${buttonStyle}

  ${({ btnStyle }) => {
    switch (btnStyle) {
      case 'danger':
        return dangerStyle;
      case 'none':
        return noneStyle;
      case 'danger-text':
        return dangerTextStyle;
      default:
        return null;
    }
  }}

  ${({ icon }) => icon && 'padding: 4px; box-shadow: none;'}

  ${({ noPadding }) => noPadding && 'padding: 0;'}
`;
