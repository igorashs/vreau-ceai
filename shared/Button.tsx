import styled, { css } from 'styled-components';

interface ButtonProps {
  btnStyle?: 'danger' | 'none' | 'danger-text' | 'accent-text' | 'dark-text';
  icon?: boolean;
  noPadding?: boolean;
}

export const buttonStyle = css`
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  text-decoration: none;

  cursor: pointer;

  :hover {
    background-color: var(--accent);
  }
`;

const noneStyle = css`
  background-color: transparent;
  box-shadow: none;

  :hover {
    background-color: #0000001a;
  }
`;

const dangerStyle = css`
  background-color: var(--danger-dark);

  :hover {
    background-color: var(--danger);
  }
`;

const textStyle = css`
  background-color: transparent;
  text-transform: initial;
  box-shadow: none;

  :hover {
    background-color: transparent;
    filter: brightness(0.7);
  }
`;

const darkTextStyle = css`
  ${textStyle}
  color: var(--text-dark);
`;

const dangerTextStyle = css`
  ${textStyle}
  color: var(--text-danger-light);
`;

const accentTextStyle = css`
  ${textStyle}
  color: var(--accent-text-dark);
`;

const Button = styled.button<ButtonProps>`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: 7px;

  ${buttonStyle}

  ${({ btnStyle }) => {
    switch (btnStyle) {
      case 'danger':
        return dangerStyle;
      case 'none':
        return noneStyle;
      case 'danger-text':
        return dangerTextStyle;
      case 'accent-text':
        return accentTextStyle;
      case 'dark-text':
        return darkTextStyle;
      default:
        return null;
    }
  }}

  ${({ icon }) => icon && 'padding: 4px; box-shadow: none;'}

  ${({ noPadding }) => noPadding && 'padding: 0;'}
`;

export default Button;
