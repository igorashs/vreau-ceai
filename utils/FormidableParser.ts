import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
