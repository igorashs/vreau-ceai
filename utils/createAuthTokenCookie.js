import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
