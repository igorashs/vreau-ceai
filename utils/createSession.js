import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
