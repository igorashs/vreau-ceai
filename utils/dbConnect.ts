import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
