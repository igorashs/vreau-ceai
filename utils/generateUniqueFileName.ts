import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
