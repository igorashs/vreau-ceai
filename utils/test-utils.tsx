import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
