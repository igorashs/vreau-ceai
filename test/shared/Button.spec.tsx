import { render } from '@/utils/test-utils';
import Button from '@/shared/Button';

describe('Button', () => {
  it('renders correctly', () => {
    const { container, getByRole, getByText } = render(
      <Button>click me</Button>,
    );

    expect(getByRole('button')).toBeInTheDocument();
    expect(getByText('click me')).toBeVisible();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('btnStyle prop', () => {
    it('renders with danger style', () => {
      const { container } = render(<Button btnStyle="danger">click me</Button>);

      expect(container.firstChild).toHaveStyleRule(
        'background-color',
        'var(--danger-dark)',
      );
      expect(container.firstChild).toHaveStyleRule(
        'background-color',
        'var(--danger)',
        {
          modifier: ':hover',
        },
      );
    });

    it('renders with none style', () => {
      const { container } = render(<Button btnStyle="none">click me</Button>);

      expect(container.firstChild).toHaveStyleRule(
        'background-color',
        'transparent',
      );
      expect(container.firstChild).toHaveStyleRule('box-shadow', 'none');
      expect(container.firstChild).toHaveStyleRule(
        'background-color',
        '#0000001a',
        { modifier: ':hover' },
      );
    });

    it('renders with danger-text style', () => {
      const { container } = render(
        <Button btnStyle="danger-text">click me</Button>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'background-color',
        'transparent',
      );
      expect(container.firstChild).toHaveStyleRule('text-transform', 'initial');
      expect(container.firstChild).toHaveStyleRule('box-shadow', 'none');
      expect(container.firstChild).toHaveStyleRule(
        'background-color',
        'transparent',
        { modifier: ':hover' },
      );
      expect(container.firstChild).toHaveStyleRule(
        'filter',
        'brightness(0.7)',
        { modifier: ':hover' },
      );
      expect(container.firstChild).toHaveStyleRule(
        'color',
        'var(--text-danger-light)',
      );
    });

    it('renders with accent text style', () => {
      const { container } = render(
        <Button btnStyle="accent-text">click me</Button>,
      );

      expect(container.firstChild).toHaveStyleRule(
        'color',
        'var(--accent-text-dark)',
      );
    });

    it('renders with dark text style', () => {
      const { container } = render(
        <Button btnStyle="dark-text">click me</Button>,
      );

      expect(container.firstChild).toHaveStyleRule('color', 'var(--text-dark)');
    });
  });

  describe('icon prop', () => {
    it('renders correctly', () => {
      const { container } = render(<Button icon>Click me</Button>);

      expect(container.firstChild).toHaveStyleRule('padding', '4px');
      expect(container.firstChild).toHaveStyleRule('box-shadow', 'none');
    });
  });

  describe('noPadding prop', () => {
    it('renders correctly', () => {
      const { container } = render(<Button noPadding>Click me</Button>);

      expect(container.firstChild).toHaveStyleRule('padding', '0');
    });
  });
});
