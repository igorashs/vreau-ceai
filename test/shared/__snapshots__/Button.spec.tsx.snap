// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c0:hover {
  background-color: var(--accent);
}

<button
  class="c0"
>
  click me
</button>
`;
